### PR TITLE
Implementation for `decimal.round` with `fractionDigits`

### PR DESCRIPTION
**Pull request by KRVPerera**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/35683

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: langlib/lang.decimal/src/main/java/org/ballerinalang/langlib/decimal/Round.java [hunk 1]
```diff
@@ -35,8 +38,18 @@
 //        isPublic = true
 //)
 public class Round {
-
-    public static BDecimal round(BDecimal x) {
-        return ValueCreator.createDecimalValue(x.value().setScale(0, RoundingMode.HALF_EVEN));
+    public static BDecimal round(BDecimal x, long fractionDigits) {
+        BigDecimal value = x.value();
+        BigDecimal scaledDecimal;
+        if (fractionDigits >= Integer.MAX_VALUE) {
```
Comment (KavinduZoysa) [inline langlib/lang.decimal/src/main/java/org/ballerinalang/langlib/decimal/Round.java:44]: Isn't it only `>`?

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibDecimalTest.java [hunk 2]
```diff
@@ -216,6 +216,60 @@ public void testRoundAsMethodInvok(BDecimal arg, String expected) {
...
+                {ValueCreator.createDecimalValue("3234.303567"), -10, "0"},
+                {ValueCreator.createDecimalValue("3234.303567"), 7, "3234.303567"},
+                {ValueCreator.createDecimalValue("3234.303567"), 10, "3234.303567"},
+                {ValueCreator.createDecimalValue("5.5555"), 4, "5.5555"},
+                {ValueCreator.createDecimalValue("5.5555"), 3, "5.556"},
+                {ValueCreator.createDecimalValue("5.5555"), 2, "5.56"},
+                {ValueCreator.createDecimalValue("5.56"), 1, "5.6"},
+                {ValueCreator.createDecimalValue("5.55"), -1, "1E+1"}
+        };
+    }
```
Comment (KavinduZoysa) [inline langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibDecimalTest.java:249]: Shall we add some tests when the x is close and x is equal to java Integer.MAX_VALUE/Integer.MIN_VALUE
just to check branches available in round function?

File: langlib/lang.decimal/src/main/java/org/ballerinalang/langlib/decimal/Round.java (hunk 1 above)
Comment (KRVPerera) [reply langlib/lang.decimal/src/main/java/org/ballerinalang/langlib/decimal/Round.java:44]: I dont think it matters.

File: langlib/lang.decimal/src/main/java/org/ballerinalang/langlib/decimal/Round.java [hunk 3]
```diff
@@ -18,13 +18,16 @@
 
 package org.ballerinalang.langlib.decimal;
 
+import io.ballerina.runtime.api.constants.RuntimeConstants;
 import io.ballerina.runtime.api.creators.ValueCreator;
 import io.ballerina.runtime.api.values.BDecimal;
 
+import java.math.BigDecimal;
+import java.math.MathContext;
```
Comment (KRVPerera) [inline langlib/lang.decimal/src/main/java/org/ballerinalang/langlib/decimal/Round.java:26]: need to remove these

File: langlib/lang.decimal/src/main/java/org/ballerinalang/langlib/decimal/Round.java [hunk 4]
```diff
@@ -35,8 +36,16 @@
 //        isPublic = true
 //)
 public class Round {
-
-    public static BDecimal round(BDecimal x) {
-        return ValueCreator.createDecimalValue(x.value().setScale(0, RoundingMode.HALF_EVEN));
+    public static BDecimal round(BDecimal x, long fractionDigits) {
+        BigDecimal value = x.value();
+        BigDecimal scaledDecimal;
+        if (fractionDigits >= value.scale()) {
```
Comment (SandaruJayawardana) [inline langlib/lang.decimal/src/main/java/org/ballerinalang/langlib/decimal/Round.java:42]: Don't we need to give an error for negative `fractionDigits` values?

Comment (gimantha) [reply langlib/lang.decimal/src/main/java/org/ballerinalang/langlib/decimal/Round.java:44]: @KavinduZoysa can we check for float case also?

Comment (KavinduZoysa) [reply langlib/lang.decimal/src/main/java/org/ballerinalang/langlib/decimal/Round.java:44]: https://github.com/ballerina-platform/ballerina-spec/issues/1074

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibDecimalTest.java (hunk 2 above)
Comment (KRVPerera) [reply langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibDecimalTest.java:259]: Thanks added by 2d77547637463307e7bfd4333e4a6d087114b7fd

File: langlib/lang.decimal/src/main/java/org/ballerinalang/langlib/decimal/Round.java (hunk 1 above)
Comment (KRVPerera) [reply langlib/lang.decimal/src/main/java/org/ballerinalang/langlib/decimal/Round.java]: Hope you agree

**Review by KavinduZoysa — DISMISSED** [review]

**Review by SandaruJayawardana — APPROVED** [review]
LGTM

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibDecimalTest.java [hunk 5]
```diff
@@ -216,6 +216,70 @@ public void testRoundAsMethodInvok(BDecimal arg, String expected) {
...
+    public void testRoundToFraction(BDecimal arg, Integer fractionDigits, String expected) {
+        Object returns = BRunUtil.invoke(compileResult, "testRoundToFractionDigits",
+                new Object[] {arg, fractionDigits});
+        assertEquals(returns, ValueCreator.createDecimalValue(expected));
+    }
+
+    @DataProvider(name = "decimalRoundToFractionDigitsValueProvider")
+    public static Object[][] decimalRoundToFractionDigitsValueProvider() {
+        return new Object[][] {
+                {ValueCreator.createDecimalValue("-504023030303030303030.303567"), 3, "-504023030303030303030.304"},
```
Comment (MaryamZi) [inline langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibDecimalTest.java:229]: I would write these tests in b7a. Creating the values in the Java side ties these tests to jBallerina and even if required we can't directly reuse the tests.

Comment (MaryamZi) [reply langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibDecimalTest.java:229]: We're also relying on the `ValueCreator` more than how a normal user would use it.

File: langlib/langlib-test/src/test/resources/test-src/decimallib_test.bal [hunk 6]
```diff
@@ -45,6 +45,18 @@ function testRound(decimal arg) returns decimal {
     return decimals:round(arg);
 }
```
Comment (MaryamZi) [inline langlib/langlib-test/src/test/resources/test-src/decimallib_test.bal:47]: Shall we also add tests with named arguments to round? That way we test the parameter names also.

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibDecimalTest.java (hunk 5 above)
Comment (KRVPerera) [reply langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibDecimalTest.java:229]: I will send a separate PR. Not sure if I can meet the deadline with this change.

Comment (gimantha) [reply langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibDecimalTest.java:229]: Lets address them in https://github.com/ballerina-platform/ballerina-lang/issues/35756

File: langlib/langlib-test/src/test/resources/test-src/decimallib_test.bal (hunk 6 above)
Comment (gimantha) [reply langlib/langlib-test/src/test/resources/test-src/decimallib_test.bal:47]: Lets address them in https://github.com/ballerina-platform/ballerina-lang/issues/35756

**Pull request merged**